README.md
# LiteNinja-Time
A library of Time-related classes for Unity3D

## Features

### Duration
A static class that can parse a string to TimeSpan and vice versa.
Units are:
- nanoseconds: ns
- microseconds: us or µs
- milliseconds: ms
- seconds: s
- minutes: m
- hours: h
- days: d
- weeks: w

Example:
```
1.5s 
1h30m
24h
1d
```

#### Parse
Parse a string into a TimeSpan. 
```
var timespan = Duration.Parse("90m");
timespan = Duration.Parse("1h30m");
timespan = Duration.Parse("1.5h");
```
If the string is invalid, an exception will be thrown.

#### ParseSafe
`ParseSafe` try to parse even if the string is not in the correct format.
```
var timespan = Duration.ParseSafe("90m1parsec");
Debug.Log(timespan.ToDuration(); // "1h30m"
```

#### TryParse
Try to parse a string into a TimeSpan. 
```
Duration.TryParse("90m", out timespan); // true, timespan = "1h30m"
Duration.TryParse("1h30m", out timespan); // true, timespan = "1h30m"
Duration.TryParse("1.5h", out timespan); // true, timespan = "1.5h"
Duration.TryParse("1h30m1parsec", out timespan); // false, timespan = "0"
```


#### Parseable
You can check if the string is in the correct format by using the `Parseable` method.

#### ToDuration
Convert a Duration to a string. The string can be in the format `{weeks}w{days}d{hours}h{minutes}m{seconds}s{milliseconds}ms`, with only the units with a value being included.
```
var timespan = Duration.Parse("90m");
Debug.Log(timespan.ToDuration()); // "1h30m"
```

#### Tests
A good coverage of the Duration class is provided by the tests. 
Peek at them to see how you can use the class.

## Install through [Unity Package Manager](https://docs.unity3d.com/Manual/upm-ui-giturl.html)

Unity's own Package Manager supports [importing packages through a URL to a Git repo](https://docs.unity3d.com/Manual/upm-ui-giturl.html):

1. First, on this repository page, click the "Clone or download" button, and copy over this repository's HTTPS URL.
2. Then click on the + button on the upper-left-hand corner of the Package Manager, select "Add package from git URL..." on the context menu, then paste this repo's URL!


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

# [1.0.2] - 2022-07-17
### Changed
- Replace the regex parsing with a more robust one based on a tokenizer.

## [1.0.1] - 2022-07-17
### Changed
- Duration now is a static class that can parse a string to TimeSpan and vice versa.

## [1.0.0] - 2022-07-16
### Added
- Duration class and extension methods

Tests/DurationTests.cs
using System;
using System.Collections;
using System.Collections.Generic;
using LiteNinja.TimeUtils.Durations;
using NUnit.Framework;
using UnityEngine;
using UnityEngine.TestTools;

namespace LiteNinja.TimeUtils.Tests
{
    public class DurationTests
    {
        private const double Millisecond = 1;
        private const double Second = 1000 * Millisecond;
        private const double Minute = 60 * Second;
        private const double Hour = 60 * Minute;
        private const double Day = 24 * Hour;
        private const double Week = 7 * Day;


        //Test Duration.Parse to check if creates correct TimeSpan
        [Test]
        public void ToTimeSpan()
        {
            var timeSpan = Duration.Parse("1s");
            Assert.AreEqual(timeSpan.TotalSeconds, 1);
            
            timeSpan = Duration.Parse("1m");
            Assert.AreEqual(timeSpan.TotalMinutes, 1);
            
            timeSpan = Duration.Parse("1h");
            Assert.AreEqual(timeSpan.TotalHours, 1);
            
            timeSpan = Duration.Parse("1.5h");
            Assert.AreEqual(timeSpan.TotalHours, 1.5);

        }

        [Test]
        public void FromTimeSpan()
        {
            var timeSpan = new TimeSpan(0, 0, 1);
            Assert.AreEqual( "1s", timeSpan.ToDuration());
            
            timeSpan = new TimeSpan(0, 1, 0);
            Assert.AreEqual( "1m", timeSpan.ToDuration());
            
            timeSpan = new TimeSpan(1, 0, 0);
            Assert.AreEqual( "1h", timeSpan.ToDuration());
            
            timeSpan = new TimeSpan(1, 30, 0);
            Assert.AreEqual( "1h30m", timeSpan.ToDuration());
        }
        
        
    }
}

Runtime/Durations/Tokenizer.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace LiteNinja.TimeUtils.Durations
{
    internal enum TokenType
    {
        Number,
        Unit,
        None
    }
    
    internal struct Token
    {
        public TokenType Type;
        public string Value;
    }
    
    internal class Tokenizer
    {
        private int _position;
        private readonly string _input;

        public Tokenizer(string input)
        {
            _input = Clean(input);
            _position = 0;
            Tokens = new List<Token>();
        }
        
        public List<Token> Tokens { get; }
        
        public List<Token> CleanTokens => Tokens.FindAll(t => t.Type != TokenType.None);
        
        public bool HasNonToken => Tokens.FindAll(t => t.Type == TokenType.None).Count > 0;

        public void Tokenize()
        {
            while (_position < _input.Length)
            {
                var token = ReadToken();
                Tokens.Add(token);
            }
        }

        private Token ReadToken()
        {
            var token = new Token
            {
                Type = TokenType.None,
                Value = string.Empty
            };
            var c = _input[_position];
            if (IsNumber(c))
            {
                token.Type = TokenType.Number;
                token.Value = ReadNumber();
            }
            else if (char.IsLetter(c))
            {
                token.Type = TokenType.Unit;
                token.Value = ReadUnit();
            }
            else
            {
                _position++;
            }
            
            return token;
        }
        
        private static bool IsNumber(char c)
        {
            return char.IsDigit(c) || c == '.';
        }
        private string ReadNumber()
        {
            return Read(IsNumber);
        }
        
        private string ReadUnit()
        {
            return Read(char.IsLetter);
        }
        
        private string Read(Func<char, bool> predicate)
        {
            var sb = new StringBuilder();
            while (_position < _input.Length)
            {
                var c = _input[_position];
                if (!predicate(c))
                {
                    break;
                }
                sb.Append(c);
                _position++;
            }
            return sb.ToString();
        }

        public void Reset()
        {
            _position = 0;
            Tokens.Clear();
        }
        
        private static string Clean(string input)
        {
            return input.Trim().ToLower().Replace(" ", "");
        }

    }
    
    
}

Runtime/Durations/DurationExtensions.cs
using System;

namespace LiteNinja.TimeUtils
{
    public static class DurationExtensions
    {
        public static Duration Negate(this Duration duration)
        {
            return new Duration(-duration.Milliseconds);
        }

        public static Duration Abs(this Duration duration)
        {
            return new Duration(Math.Abs(duration.Milliseconds));
        }

        public static DateTime After(this Duration self, DateTime dateTime)
        {
            return dateTime.AddMilliseconds(self.Milliseconds);
        }


        public static DateTime Before(this Duration self, DateTime dateTime)
        {
            return dateTime.AddMilliseconds(-self.Milliseconds);
        }

        public static DateTime Before(this DateTime self, Duration duration)
        {
            return self.AddMilliseconds(-duration.Milliseconds);
        }

        public static DateTime After(this DateTime self, Duration duration)
        {
            return self.AddMilliseconds(duration.Milliseconds);
        }

        public static Duration Since(this DateTime self)
        {
            return new Duration(DateTime.Now.Subtract(self).TotalMilliseconds);
        }

        public static Duration Until(this DateTime self)
        {
            return new Duration(self.Subtract(DateTime.Now).TotalMilliseconds);
        }

        public static Duration Between(this DateTime self, DateTime other)
        {
            return new Duration(self.Subtract(other).TotalMilliseconds);
        }
        
        public static TimeSpan ToTimeSpan(this Duration self)
        {
            return TimeSpan.FromMilliseconds(self.Milliseconds);
        }
        
        public static Duration ToDuration(this TimeSpan self)
        {
            return new Duration(self.TotalMilliseconds);
        }
        
        
    }
}

Runtime/Durations/Duration.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace LiteNinja.TimeUtils.Durations
{
    public static class Duration
    {
        #region Consts
        private const double Nanosecond = Microsecond / 1000;
        private const double Microsecond = Millisecond / 1000;
        private const double Millisecond = 1;
        private const double Second = 1000 * Millisecond;
        private const double Minute = 60 * Second;
        private const double Hour = 60 * Minute;
        private const double Day = 24 * Hour;
        private const double Week = 7 * Day;

        private static readonly Dictionary<string, double> UnitToMillisecond = new()
        {
            { "ns", Nanosecond },
            { "us", Microsecond },
            { "µs", Microsecond },
            { "ms", Millisecond },
            { "s", Second },
            { "m", Minute },
            { "h", Hour },
            { "d", Day },
            { "w", Week }
        };
        #endregion

        /// <summary>
        /// Convert a duration string to a time span, even if it's not a valid duration string.
        /// </summary>
        public static TimeSpan SafeParse(string duration)
        {
            return Parse(duration);
        }
        
        /// <summary>
        /// Convert a duration string to a time span.
        /// if ignoreInvalid is true, the part of the string that is not a valid duration will be ignored.
        /// </summary>
        public static TimeSpan Parse(string duration, bool ignoreInvalid = false)
        {
            var tokenizer = new Tokenizer(duration);
            tokenizer.Tokenize();
            var tokens = ignoreInvalid ? tokenizer.CleanTokens : tokenizer.Tokens;
            double totalMilliseconds = 0;
            if (tokens.Count == 0) return new TimeSpan(0);
            var index = 0;
            while (index < tokens.Count)
            {
                var token = tokens[index];
                if (token.Type == TokenType.Number)
                {
                    var number = double.Parse(token.Value);
                    index = ExtractUnit(duration, ignoreInvalid, index, tokens, number, ref totalMilliseconds);
                }
                else
                {
                    index++;
                }
            }
            
            return TimeSpan.FromMilliseconds(totalMilliseconds);
        }

        

        /// <summary>
        /// Check if a string is a completely valid duration string.
        /// </summary>
        public static bool Parseable(string duration)
        {
            var tokenizer = new Tokenizer(duration);
            tokenizer.Tokenize();
            return !tokenizer.HasNonToken;
        }
        
        
        
        /// <summary>
        /// Try to parse a duration string to a time span.
        /// </summary>
        public static bool TryParse(string duration, out TimeSpan timeSpan)
        {
            try
            {
                timeSpan = Parse(duration);
                return true;
            }
            catch
            {
                timeSpan = new TimeSpan();
                return false;
            }
        }

        /// <summary>
        /// Convert a TimeSpan to a duration string.
        /// </summary>
        public static string ToDuration(this TimeSpan timeSpan)
        {
            var result = "";
            var milliseconds = timeSpan.TotalMilliseconds;
            var sign = milliseconds < 0 ? "-" : "";
            if (milliseconds == 0)
            {
                return "0";
            }

            // Weeks
            var weeks = Math.Floor(milliseconds / Week);
            if (weeks > 0)
            {
                result += $"{weeks}w";
                milliseconds -= weeks * Week;
            }

            // Days
            var days = Math.Floor(milliseconds / Day);
            if (days > 0)
            {
                result += $"{days}d";
                milliseconds -= days * Day;
            }

            //hours
            var hours = Math.Floor(milliseconds / Hour);
            if (hours > 0)
            {
                result += $"{hours}h";
                milliseconds -= hours * Hour;
            }

            //minutes
            var minutes = Math.Floor(milliseconds / Minute);
            if (minutes > 0)
            {
                result += $"{minutes}m";
                milliseconds -= minutes * Minute;
            }

            //seconds
            var seconds = Math.Floor(milliseconds / Second);
            if (seconds > 0)
            {
                result += $"{seconds}s";
                milliseconds -= seconds * Second;
            }

            //milliseconds
            var floorMilliseconds = Math.Floor(milliseconds);
            if (floorMilliseconds > 0)
            {
                result += $"{floorMilliseconds}ms";
                milliseconds -= floorMilliseconds;
            }
            
            //microseconds
            var microseconds = Math.Floor(milliseconds / Microsecond);
            if (microseconds > 0)
            {
                result += $"{microseconds}µs";
                milliseconds -= microseconds * Microsecond;
            }
            
            //nanoseconds
            var nanoseconds = Math.Floor(milliseconds / Nanosecond);
            if (nanoseconds > 0)
            {
                result += $"{nanoseconds}ns";
                milliseconds -= nanoseconds * Nanosecond;
            }
            
            return $"{sign}{result}";
        }
        
        #region Private methods
        private static int ExtractUnit(string duration, bool ignoreInvalid, int index, IReadOnlyList<Token> tokens, double number,
            ref double totalMilliseconds)
        {
            if (index + 1 < tokens.Count)
            {
                var unitToken = tokens[index + 1];
                if (unitToken.Type == TokenType.Unit)
                {
                    totalMilliseconds = AddMilliseconds(ignoreInvalid, unitToken, totalMilliseconds, number,
                        ref index);
                }
                else
                {
                    throw new FormatException($"Invalid duration format: {duration}");
                }
            }
            else
            {
                totalMilliseconds += number;
                index++;
            }

            return index;
        }

        private static double AddMilliseconds(bool ignoreInvalid, Token unitToken, double totalMilliseconds, double number,
            ref int index)
        {
            var unit = unitToken.Value;
            if (UnitToMillisecond.TryGetValue(unit, out var unitMilliseconds))
            {
                totalMilliseconds += number * unitMilliseconds;
                index += 2;
            }
            else
            {
                if (ignoreInvalid)
                {
                    index += 2;
                }
                else
                {
                    throw new ArgumentException($"Invalid unit: {unit}");
                }
            }

            return totalMilliseconds;
        }
        #endregion
        
    }
}